src/blas/impl/KokkosBlas_util.hpp

```diff
@@ -45,7 +45,121 @@
 #ifndef KOKKOS_BLAS_UTIL_HPP
 #define KOKKOS_BLAS_UTIL_HPP
 
+#include "Kokkos_ArithTraits.hpp"
+
+// TPL macros
+#if defined(KOKKOSKERNELS_ENABLE_TPL_MKL)
+#include "mkl_version.h"
+#if __INTEL_MKL__ >= 2018
+#define __KOKKOSBLAS_ENABLE_INTEL_MKL_COMPACT__ 1
+#include "mkl_compact.h"
+#endif
+#endif
+
 namespace KokkosBlas {
+
+//////// Tags for BLAS ////////
+struct Trans {
+  struct Transpose {};
+  struct NoTranspose {};
+  struct ConjTranspose {};
+};
+
+struct Algo {
+  struct Level3 {
+    struct Unblocked {
+      static const char *name() { return "Unblocked"; }
+    };
+    struct Blocked {
+      static const char *name() { return "Blocked"; }
+      // TODO:: for now harwire the blocksizes; this should reflect
+      // regieter blocking (not about team parallelism).
+      // this mb should vary according to
+      // - team policy (smaller) or range policy (bigger)
+      // - space (gpu vs host)
+      // - blocksize input (blk <= 4 mb = 2, otherwise mb = 4), etc.
+#if defined(KOKKOS_IF_ON_HOST)
```

Comment:
Thank you for pointing that out: I moved whole thing from [src/batched/KokkosBatched_Util.hpp](https://github.com/kokkos/kokkos-kernels/pull/1433/files#diff-e121dfec07f62547224c7a5255ecac13bfa6566b1c43af6ea55c75f78b2599b1L354-L447) and missed `algo_level{2,3}_blocked_mb_impl` (fixed that on https://github.com/kokkos/kokkos-kernels/pull/1433/commits/24db95bfe59bf4c543c8f682352216e7899471a6 for consistency).

If I understand correctly, @dalg24 replaced on https://github.com/kokkos/kokkos-kernels/pull/1265 the old approach based on `Kokkos::Impl::ActiveExecutionMemorySpace` with new one based on `KOKKOS_IF_ON_{HOST,DEVICE}` - which corresponds to the former getting deprecated with https://github.com/kokkos/kokkos/pull/4660.

Since we probably can't get back to deprecated `Kokkos::Impl::ActiveExecutionMemorySpace`, we could consider passing actual execution space type through the algorithm (e.g. `gemv` → ... → `SerialGemvInternal`) and template `mb()` on it (same way like `algo_level{2,3}_blocked_mb_impl` was templated on execution space to yield proper block sizes for CPU and GPU) to be called like:
```cpp
constexpr int mbAlgo = Algo::Gemv::Blocked::mb<ExecSpace>();
```

Reply:
Can this be addressed on new PR for passing execution space ?
Please see [KokkosBlas_util.hpp](https://github.com/kokkos/kokkos-kernels/pull/1489/files#diff-766b8d7f414bd1fe2b8df07e69ea2ce1a9512dc790dd65d6818b605215c86c2cR74-R125) on #1489, which aims to pass execution space as explicit argument in the algorithms. Based on that, we can template `mb()` on execution space type and calculate block size accordingly - without using `KOKKOS_IF_ON_{HOST,DEVICE}`.

Proposed approach moves `mb()` out of `Algo::Level#::Blocked`, so we don't need to template `Blocked` on `ExecSpace` too (i.e. keep it as a plain tag). It seems to fit the overall idea of passing execution space instance better, while templating `Blocked` on execution space would yield less flexible architecture with the advantage of smaller and simpler changes (leaving `Unblocked` variants intact).

Reply:
Yes I view it as an improvement not a requirement for this PR so doing it in a later PR is fine